### PR TITLE
CRX Manifest Comment Update and Exception Handling

### PR DESCRIPTION
**Pull request by jaredswilson**
- Added updated functionality to remove comments from CRX manifests before parsing.
 - Added further exception handling for parameter passing and password protected files.

### CODE REVIEW

File: permhash/helpers.py [hunk 1]
```diff
@@ -76,60 +81,73 @@ def check_type(path, mime):
...
+                                newpermkeys
+                                + "."
+                                + str(embeddedpermkeys)
+                                + "."
+                                + str(embeddedpermvalues)
+                            )
+            return newpermlist
+        else:
+            logging.info(
+                "There are no standard permissions in this Manifest: %s.", path
```
Comment (jconnor0426) [inline permhash/helpers.py]: It looks like path is only being used in this function for this log statement. 

I'd recommend keeping the path variable out of the function and moving the logging around the path failing to the calling function. This keeps the function a  little more focused

File: permhash/helpers.py [hunk 2]
```diff
@@ -179,17 +206,26 @@ def create_crx_permlist(path):
...
+            manifest_json = strip_comments(manifest_text, path)
+        else:
             try:
-                manifest_json = json.loads(manifest_text.encode().decode("utf-8-sig"))
-            except JSONDecodeError:
-                logging.warning(
-                    "This Manifest file is abnormal and unable to be read: %s.", path
-                )
+                manifest_json = json.loads(manifest_text)
+            except OSError:
```
Comment (jconnor0426) [inline permhash/helpers.py]: It looks like OSError isn't going to ever get triggered since the json loads is operating on a string not a file. 

Can we remove it?

File: permhash/helpers.py [hunk 3]
```diff
@@ -179,17 +206,26 @@ def create_crx_permlist(path):
...
+                    manifest_json = json.loads(
+                        manifest_text.encode().decode("utf-8-sig")
+                    )
+                except JSONDecodeError:
+                    logging.warning(
+                        "This manifest file is abnormal and unable to be read: %s.",
+                        path,
+                    )
+                    return False
+        perm_list = parse_crx_manifest(manifest_json, path)
```
Comment (jconnor0426) [inline permhash/helpers.py]: Related to the comment on path, I'd recommend adding a check here to see if perm_list is False here and log that there was an error

File: permhash/helpers.py [hunk 4]
```diff
@@ -205,11 +241,40 @@ def create_crx_manifest_permlist(path):
     if check_type(path, CRX_MANIFEST_MIMETYPES):
         with open(path, encoding="utf-8") as manifest:
             try:
-                manifest_json = json.load(manifest)
```
Comment (jconnor0426) [inline permhash/helpers.py:208]: Dealing with the text encodings is getting pretty complicated and it looks like this code is largely the same as what you see in `create_crx_permlist`. Updating it in the future to deal with new possible encoding issues could be problematic.

My recommendation would be to have both `create_crx_manifest_permlist` and `create_crx_permlist` to get the manifest file into a bytestring (don't mess with encodings) and then  call a function called `process_manifest_bytes` which takes the bytestring and tries to decode and do all the json processing. Then both `create_crx_manifest_permlist` and `create_crx_permlist`  can call parse_crx_manifest as normal

File: permhash/helpers.py [hunk 5]
```diff
@@ -314,3 +379,31 @@ def create_apk_permlist(path):
...
+
+    :param manifest_text: A read manifest file from input
+    :type manifest_text: read object
+    :param path: A path to the manifest
+    :type path: string
+    """
+    try:
+        stripped_manifest = jstyleson.loads(manifest_text)
+        return stripped_manifest
+    except OSError:
```
Comment (jconnor0426) [inline permhash/helpers.py]: Not going to need OSerror since the text should already be read from a file

File: permhash/helpers.py [hunk 6]
```diff
@@ -205,11 +241,40 @@ def create_crx_manifest_permlist(path):
...
+                    logging.warning("This manifest is unable to be read: %s.", path)
+                    return False
+            if (
+                "//" in manifest_text or "/*" in manifest_text
+            ):  # Enter to strip comments if there is suspected comments in the text.
+                manifest_json = strip_comments(manifest_text, path)
+            else:
+                try:
+                    manifest_json = json.loads(manifest_text)
+                except OSError:
```
Comment (jconnor0426) [inline permhash/helpers.py]: Same here no chance for loads to throw an OSError (https://docs.python.org/3/library/json.html)

**Review by jconnor0426 — CHANGES_REQUESTED** [review]
One big suggestion to consolidate all the work to go from manifest bytes to python objects into a single function. 

Having one place to deal with all the encoding and comment issues will be helpful for long term maintenance

File: permhash/helpers.py (hunk 1 above)
Comment (jaredswilson) [reply permhash/helpers.py]: Agreed. Updated the logging and moved the path to the calling function.

File: permhash/helpers.py (hunk 2 above)
Comment (jaredswilson) [reply permhash/helpers.py]: Yes, this is more relevant in the APK handling but makes sense to not be in these cases.

File: permhash/helpers.py (hunk 4 above)
Comment (jaredswilson) [reply permhash/helpers.py:208]: Agree. I was writing each part and they seemed to be coming together in a _different way_ but are handling the same data. Function created. Will be posted soon.

File: permhash/helpers.py (hunk 3 above)
Comment (jaredswilson) [reply permhash/helpers.py]: That's exactly how it was handled with moving the path. We check if it is false and raise as needed.

File: permhash/helpers.py [hunk 7]
```diff
@@ -199,18 +219,29 @@ def create_crx_manifest_permlist(path):
...
+                try:
+                    manifest_byte_read = manifest_byte_stream.read()
+                except IOError as error:
+                    # This file was unable to be read
+                    logging.warning("This manifest is unable to be read.")
+                    logging.warning(str(error))
+                    return False
+                processed_json = process_manifest_bytes(manifest_byte_read)
+                if not processed_json:
+                    logging.warning("Path: %s", path)
```
Comment (jconnor0426) [inline permhash/helpers.py:264]: Should probably return here right? Otherwise we'll pass the value false into `parse_crx_manifest`

**Review by jconnor0426 — CHANGES_REQUESTED** [review]
One small comment and here and a request to add some tests to validate the appropriate behavior on the new edge cases. I.E.

Test: 
* Run parse manifest on a directory
* Run parse manifest on manifest files with encodings that we've seen
* Run parse manifest on a manifest file with json comments as we've seen

Any other edgecases this PR is supposed to be covering

**Comment by jaredswilson** [comment]
@jconnor0426 this most recent update should have all the proper modifications and tests.
